File: apps/ios/ConstructionManager/ConstructionManager/Features/DailyLogs/Data/DailyLogsRepository.swift
//
//  DailyLogsRepository.swift
//  ConstructionManager
//
//  Repository for Daily Logs feature with GraphQL and offline support
//

import Foundation
import Combine

// MARK: - Result Types

enum DailyLogsResult<T> {
    case success(T)
    case error(String, isOffline: Bool)
    case loading
}

// MARK: - Models

struct DailyLogSummary: Identifiable, Codable {
    let id: String
    let date: Date
    let status: String
    let notes: String?
    let projectId: String
    let projectName: String
    let submitterName: String
    let photoCount: Int
    let crewCount: Int
    let entriesCount: Int
    let totalLaborHours: Double
    let isPending: Bool

    init(
        id: String,
        date: Date,
        status: String,
        notes: String?,
        projectId: String,
        projectName: String,
        submitterName: String,
        photoCount: Int = 0,
        crewCount: Int = 0,
        entriesCount: Int = 0,
        totalLaborHours: Double = 0,
        isPending: Bool = false
    ) {
        self.id = id
        self.date = date
        self.status = status
        self.notes = notes
        self.projectId = projectId
        self.projectName = projectName
        self.submitterName = submitterName
        self.photoCount = photoCount
        self.crewCount = crewCount
        self.entriesCount = entriesCount
        self.totalLaborHours = totalLaborHours
        self.isPending = isPending
    }
}

struct DailyLogDetailModel: Identifiable, Codable {
    let id: String
    let date: Date
    let status: String
    let notes: String?
    let weatherDelay: Bool
    let weatherDelayNotes: String?
    let gpsLatitude: Double?
    let gpsLongitude: Double?
    let projectId: String
    let projectName: String
    let projectAddress: String?
    let submitterId: String
    let submitterName: String
    let submitterEmail: String?
    let approverId: String?
    let approverName: String?
    let photos: [DailyLogPhotoModel]
    let entries: [DailyLogEntryModel]
    let crewMembers: [CrewMemberModel]
    let materials: [MaterialModel]
    let issues: [IssueModel]
    let photoCount: Int
    let crewCount: Int
    let entriesCount: Int
    let totalLaborHours: Double
    let createdAt: Date
    let updatedAt: Date
}

struct DailyLogPhotoModel: Identifiable, Codable {
    let id: String
    let url: String
    let caption: String?
    let gpsLatitude: Double?
    let gpsLongitude: Double?
}

struct DailyLogEntryModel: Identifiable, Codable {
    let id: String
    let activityName: String
    let locationNames: [String]
    let statusName: String?
    let percentComplete: Int?
    let notes: String?
}

struct CrewMemberModel: Identifiable, Codable {
    let id: String
    let name: String
    let hours: Double
    let trade: String?
}

struct MaterialModel: Identifiable, Codable {
    let id: String
    let name: String
    let quantity: Double
    let unit: String?
    let notes: String?
}

struct IssueModel: Identifiable, Codable {
    let id: String
    let description: String
    let severity: String
    let resolved: Bool
}

// MARK: - Repository

@MainActor
class DailyLogsRepository: ObservableObject {
    static let shared = DailyLogsRepository()

    @Published private(set) var dailyLogs: [DailyLogSummary] = []
    @Published private(set) var isLoading = false
    @Published private(set) var isUsingCachedData = false
    @Published var error: String?

    private let graphQLClient = GraphQLClient.shared
    private let networkMonitor = NetworkMonitor.shared
    private let syncQueue = SyncQueue.shared

    private var fetchTask: Task<Void, Never>?
    private var lastFetchTime: Date?
    private let minFetchInterval: TimeInterval = 2.0

    private init() {}

    // MARK: - Fetch Daily Logs

    func fetchDailyLogs(
        projectId: String? = nil,
        status: String? = nil,
        startDate: Date? = nil,
        endDate: Date? = nil,
        forceRefresh: Bool = false
    ) async {
        // Skip if recently fetched
        if !forceRefresh, let lastFetch = lastFetchTime,
           Date().timeIntervalSince(lastFetch) < minFetchInterval {
            return
        }

        // Deduplicate requests
        if let existingTask = fetchTask {
            await existingTask.value
            return
        }

        fetchTask = Task {
            await performFetch(
                projectId: projectId,
                status: status,
                startDate: startDate,
                endDate: endDate,
                forceRefresh: forceRefresh
            )
        }

        await fetchTask?.value
        fetchTask = nil
    }

    private func performFetch(
        projectId: String?,
        status: String?,
        startDate: Date?,
        endDate: Date?,
        forceRefresh: Bool
    ) async {
        isLoading = true
        error = nil
        isUsingCachedData = false
        defer { isLoading = false }

        // Check network connectivity
        guard networkMonitor.isConnected else {
            loadFromOfflineCache(projectId: projectId)
            return
        }

        // TODO: Use GraphQL once code generation is complete
        // let query = GetDailyLogsQuery(
        //     projectId: projectId.map { .some($0) } ?? .null,
        //     status: status.map { .some(DailyLogStatus(rawValue: $0) ?? .draft) } ?? .null,
        //     startDate: startDate.map { .some(ISO8601DateFormatter().string(from: $0)) } ?? .null,
        //     endDate: endDate.map { .some(ISO8601DateFormatter().string(from: $0)) } ?? .null,
        //     page: .some(1),
        //     pageSize: .some(50)
        // )

        // Fallback to REST
        await fetchDailyLogsViaREST(projectId: projectId, status: status)
    }

    /// Fallback to REST API
    private func fetchDailyLogsViaREST(projectId: String?, status: String?) async {
        do {
            var queryItems: [URLQueryItem] = []
            if let projectId = projectId {
                queryItems.append(URLQueryItem(name: "projectId", value: projectId))
            }
            if let status = status {
                queryItems.append(URLQueryItem(name: "status", value: status))
            }

            let response: DailyLogsResponse = try await APIClient.shared.get(
                "/daily-logs",
                queryItems: queryItems.isEmpty ? nil : queryItems
            )

            self.dailyLogs = response.dailyLogs.map { apiLog in
                DailyLogSummary(
                    id: apiLog.id,
                    date: apiLog.date,
                    status: apiLog.status,
                    notes: apiLog.notes,
                    projectId: apiLog.projectId,
                    projectName: apiLog.projectName ?? "Unknown Project",
                    submitterName: apiLog.submitterName ?? "Unknown",
                    photoCount: 0,
                    crewCount: apiLog.crewCount,
                    entriesCount: apiLog.entriesCount ?? 0,
                    totalLaborHours: apiLog.totalHours,
                    isPending: false
                )
            }
            self.lastFetchTime = Date()

            // Cache for offline use
            cacheDailyLogs(dailyLogs, projectId: projectId)

        } catch {
            self.error = error.localizedDescription
            loadFromOfflineCache(projectId: projectId)
        }
    }

    // MARK: - Get Single Daily Log

    func getDailyLog(id: String) async -> DailyLogsResult<DailyLogDetailModel> {
        guard networkMonitor.isConnected else {
            // TODO: Load from cache
            return .error("No internet connection", isOffline: true)
        }

        // TODO: Use GraphQL once code generation is complete
        // let query = GetDailyLogQuery(id: id)

        do {
            let response: DailyLogResponse = try await APIClient.shared.get("/daily-logs/\(id)")
            let apiLog = response.dailyLog

            let detail = DailyLogDetailModel(
                id: apiLog.id,
                date: apiLog.date,
                status: apiLog.status,
                notes: apiLog.notes,
                weatherDelay: apiLog.weatherDelay,
                weatherDelayNotes: apiLog.weatherDelayNotes,
                gpsLatitude: nil,
                gpsLongitude: nil,
                projectId: apiLog.projectId,
                projectName: apiLog.projectName ?? "Unknown Project",
                projectAddress: nil,
                submitterId: apiLog.submittedBy ?? "",
                submitterName: apiLog.submitterName ?? "Unknown",
                submitterEmail: nil,
                approverId: nil,
                approverName: nil,
                photos: [],
                entries: [],
                crewMembers: [],
                materials: [],
                issues: [],
                photoCount: 0,
                crewCount: apiLog.crewCount,
                entriesCount: apiLog.entriesCount ?? 0,
                totalLaborHours: apiLog.totalHours,
                createdAt: apiLog.createdAt,
                updatedAt: apiLog.updatedAt
            )

            return .success(detail)

        } catch {
            return .error(error.localizedDescription, isOffline: false)
        }
    }

    // MARK: - Create Daily Log

    func createDailyLog(
        projectId: String,
        date: Date,
        notes: String?,
        weatherDelay: Bool = false,
        weatherDelayNotes: String? = nil,
        gpsLatitude: Double? = nil,
        gpsLongitude: Double? = nil
    ) async -> DailyLogsResult<String> {
        let request = CreateDailyLogRequest(
            projectId: projectId,
            date: date,
            notes: notes,
            weatherDelay: weatherDelay,
            weatherDelayNotes: weatherDelayNotes,
            gpsLatitude: gpsLatitude,
            gpsLongitude: gpsLongitude
        )

        guard networkMonitor.isConnected else {
            // Queue for offline sync
            return createDailyLogOffline(request: request)
        }

        // TODO: Use GraphQL mutation
        // let mutation = CreateDailyLogMutation(input: CreateDailyLogInput(...))

        do {
            let response: DailyLogResponse = try await APIClient.shared.post(
                "/daily-logs",
                body: request
            )

            // Refresh list
            await fetchDailyLogs(projectId: projectId, forceRefresh: true)

            return .success(response.dailyLog.id)

        } catch {
            // Queue for offline sync on failure
            return createDailyLogOffline(request: request)
        }
    }

    private func createDailyLogOffline(request: CreateDailyLogRequest) -> DailyLogsResult<String> {
        let tempId = "temp_\(UUID().uuidString)"

        // Create local entry
        let localLog = DailyLogSummary(
            id: tempId,
            date: request.date,
            status: "DRAFT",
            notes: request.notes,
            projectId: request.projectId,
            projectName: "",
            submitterName: "",
            isPending: true
        )

        dailyLogs.insert(localLog, at: 0)

        // Queue for sync
        syncQueue.enqueue(
            type: .create,
            resourceType: .dailyLog,
            resourceId: tempId,
            payload: request
        )

        return .success(tempId)
    }

    // MARK: - Submit Daily Log

    func submitDailyLog(id: String) async -> DailyLogsResult<Void> {
        guard networkMonitor.isConnected else {
            // Queue for offline sync
            syncQueue.enqueue(
                type: .submit,
                resourceType: .dailyLog,
                resourceId: id,
                payload: ["id": id]
            )
            return .success(())
        }

        // TODO: Use GraphQL mutation
        // let mutation = SubmitDailyLogMutation(id: id)

        do {
            let _: DailyLogResponse = try await APIClient.shared.post(
                "/daily-logs/\(id)/submit",
                body: EmptyBody()
            )
            return .success(())
        } catch {
            return .error(error.localizedDescription, isOffline: false)
        }
    }

    // MARK: - Offline Cache

    private func loadFromOfflineCache(projectId: String?) {
        let cached = loadCachedDailyLogs(projectId: projectId)
        if !cached.isEmpty {
            self.dailyLogs = cached
            self.isUsingCachedData = true
        } else {
            self.error = "No internet connection and no cached data available"
        }
    }

    private func cacheDailyLogs(_ logs: [DailyLogSummary], projectId: String?) {
        guard let data = try? JSONEncoder().encode(logs) else { return }

        let fileManager = FileManager.default
        guard let cachesDir = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first else { return }

        let filename = projectId != nil ? "cached_daily_logs_\(projectId!).json" : "cached_daily_logs.json"
        let cacheFile = cachesDir.appendingPathComponent(filename)
        try? data.write(to: cacheFile)
    }

    private func loadCachedDailyLogs(projectId: String?) -> [DailyLogSummary] {
        let fileManager = FileManager.default
        guard let cachesDir = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first else { return [] }

        let filename = projectId != nil ? "cached_daily_logs_\(projectId!).json" : "cached_daily_logs.json"
        let cacheFile = cachesDir.appendingPathComponent(filename)

        guard let data = try? Data(contentsOf: cacheFile),
              let logs = try? JSONDecoder().decode([DailyLogSummary].self, from: data) else {
            return []
        }

        return logs
    }
}

// MARK: - Request Models

private struct EmptyBody: Encodable {}

File: apps/android/construction-android/app/src/main/java/com/constructionpro/app/features/dailylogs/data/DailyLogsRepository.kt
package com.constructionpro.app.features.dailylogs.data

import com.apollographql.apollo3.ApolloClient
import com.apollographql.apollo3.api.Optional
import com.apollographql.apollo3.cache.normalized.FetchPolicy
import com.apollographql.apollo3.cache.normalized.fetchPolicy
import com.constructionpro.app.data.local.DailyLogDao
import com.constructionpro.app.data.local.DailyLogEntity
import com.constructionpro.app.data.local.PendingActionDao
import com.constructionpro.app.data.local.PendingActionEntity
import com.constructionpro.app.graphql.CreateDailyLogMutation
import com.constructionpro.app.graphql.GetDailyLogQuery
import com.constructionpro.app.graphql.GetDailyLogsQuery
import com.constructionpro.app.graphql.SubmitDailyLogMutation
import com.constructionpro.app.graphql.UpdateDailyLogMutation
import com.constructionpro.app.graphql.type.CreateDailyLogInput
import com.constructionpro.app.graphql.type.DailyLogStatus
import com.constructionpro.app.graphql.type.UpdateDailyLogInput
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.time.LocalDate
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

sealed class DailyLogsResult<out T> {
    data class Success<T>(val data: T) : DailyLogsResult<T>()
    data class Error(val message: String, val isOffline: Boolean = false) : DailyLogsResult<Nothing>()
    data object Loading : DailyLogsResult<Nothing>()
}

data class DailyLogSummary(
    val id: String,
    val date: LocalDate,
    val status: String,
    val notes: String?,
    val projectId: String,
    val projectName: String,
    val submitterName: String,
    val photoCount: Int,
    val crewCount: Int,
    val entriesCount: Int,
    val isPending: Boolean = false
)

data class DailyLogDetail(
    val id: String,
    val date: LocalDate,
    val status: String,
    val notes: String?,
    val weatherDelay: Boolean,
    val weatherDelayNotes: String?,
    val gpsLatitude: Double?,
    val gpsLongitude: Double?,
    val projectId: String,
    val projectName: String,
    val projectAddress: String?,
    val submitterId: String,
    val submitterName: String,
    val submitterEmail: String?,
    val approverId: String?,
    val approverName: String?,
    val photos: List<DailyLogPhoto>,
    val entries: List<DailyLogEntry>,
    val crewMembers: List<CrewMember>,
    val materials: List<Material>,
    val issues: List<Issue>,
    val photoCount: Int,
    val crewCount: Int,
    val entriesCount: Int,
    val totalLaborHours: Double
)

data class DailyLogPhoto(
    val id: String,
    val url: String,
    val caption: String?,
    val gpsLatitude: Double?,
    val gpsLongitude: Double?
)

data class DailyLogEntry(
    val id: String,
    val activityName: String,
    val locationNames: List<String>,
    val statusName: String?,
    val percentComplete: Int?,
    val notes: String?
)

data class CrewMember(
    val id: String,
    val name: String,
    val hours: Double,
    val trade: String?
)

data class Material(
    val id: String,
    val name: String,
    val quantity: Double,
    val unit: String?,
    val notes: String?
)

data class Issue(
    val id: String,
    val description: String,
    val severity: String,
    val resolved: Boolean
)

@Singleton
class DailyLogsRepository @Inject constructor(
    private val apolloClient: ApolloClient,
    private val dailyLogDao: DailyLogDao,
    private val pendingActionDao: PendingActionDao
) {
    private val json = Json { ignoreUnknownKeys = true }

    fun getDailyLogs(
        projectId: String? = null,
        status: String? = null,
        startDate: LocalDate? = null,
        endDate: LocalDate? = null,
        forceRefresh: Boolean = false
    ): Flow<DailyLogsResult<List<DailyLogSummary>>> = flow {
        emit(DailyLogsResult.Loading)

        try {
            // First emit cached data
            val cached = if (projectId != null) {
                dailyLogDao.getByProject(projectId)
            } else {
                dailyLogDao.getAll()
            }

            if (cached.isNotEmpty()) {
                emit(DailyLogsResult.Success(cached.map { it.toSummary() }))
            }

            // Then fetch from network
            val fetchPolicy = if (forceRefresh) FetchPolicy.NetworkOnly else FetchPolicy.CacheFirst

            val response = apolloClient.query(
                GetDailyLogsQuery(
                    projectId = Optional.presentIfNotNull(projectId),
                    status = Optional.presentIfNotNull(status?.let { DailyLogStatus.safeValueOf(it) }),
                    startDate = Optional.presentIfNotNull(startDate?.toString()),
                    endDate = Optional.presentIfNotNull(endDate?.toString()),
                    page = Optional.present(1),
                    pageSize = Optional.present(50)
                )
            )
                .fetchPolicy(fetchPolicy)
                .execute()

            if (response.hasErrors()) {
                val errorMessage = response.errors?.firstOrNull()?.message ?: "Unknown error"
                emit(DailyLogsResult.Error(errorMessage))
                return@flow
            }

            val logs = response.data?.dailyLogs?.edges?.map { edge ->
                val node = edge.node
                DailyLogSummary(
                    id = node.id,
                    date = LocalDate.parse(node.date.toString()),
                    status = node.status.name,
                    notes = node.notes,
                    projectId = node.project.id,
                    projectName = node.project.name,
                    submitterName = node.submitter.name,
                    photoCount = node.photoCount,
                    crewCount = node.crewCount,
                    entriesCount = node.entriesCount,
                    isPending = false
                )
            } ?: emptyList()

            // Cache the results
            val entities = logs.map { log ->
                DailyLogEntity(
                    id = log.id,
                    date = log.date.toString(),
                    status = log.status,
                    notes = log.notes,
                    projectId = log.projectId,
                    projectName = log.projectName,
                    submitterName = log.submitterName,
                    crewCount = log.crewCount,
                    totalHours = 0.0,
                    pendingSync = false,
                    updatedAt = System.currentTimeMillis()
                )
            }
            dailyLogDao.insertAll(entities)

            emit(DailyLogsResult.Success(logs))

        } catch (e: Exception) {
            // Network error - return cached data with offline flag
            val cached = if (projectId != null) {
                dailyLogDao.getByProject(projectId)
            } else {
                dailyLogDao.getAll()
            }

            if (cached.isNotEmpty()) {
                emit(DailyLogsResult.Success(cached.map { it.toSummary() }))
            } else {
                emit(DailyLogsResult.Error(e.message ?: "Network error", isOffline = true))
            }
        }
    }.flowOn(Dispatchers.IO)

    suspend fun getDailyLog(id: String): DailyLogsResult<DailyLogDetail> = withContext(Dispatchers.IO) {
        try {
            val response = apolloClient.query(GetDailyLogQuery(id))
                .fetchPolicy(FetchPolicy.CacheFirst)
                .execute()

            if (response.hasErrors()) {
                val errorMessage = response.errors?.firstOrNull()?.message ?: "Unknown error"
                return@withContext DailyLogsResult.Error(errorMessage)
            }

            val log = response.data?.dailyLog ?: return@withContext DailyLogsResult.Error("Daily log not found")

            DailyLogsResult.Success(
                DailyLogDetail(
                    id = log.id,
                    date = LocalDate.parse(log.date.toString()),
                    status = log.status.name,
                    notes = log.notes,
                    weatherDelay = log.weatherDelay,
                    weatherDelayNotes = log.weatherDelayNotes,
                    gpsLatitude = log.gpsLatitude,
                    gpsLongitude = log.gpsLongitude,
                    projectId = log.project.id,
                    projectName = log.project.name,
                    projectAddress = log.project.address.formatted,
                    submitterId = log.submitter.id,
                    submitterName = log.submitter.name,
                    submitterEmail = log.submitter.email,
                    approverId = log.approver?.id,
                    approverName = log.approver?.name,
                    photos = log.photos.map { photo ->
                        DailyLogPhoto(
                            id = photo.id,
                            url = photo.url,
                            caption = photo.caption,
                            gpsLatitude = photo.gpsLatitude,
                            gpsLongitude = photo.gpsLongitude
                        )
                    },
                    entries = log.entries.map { entry ->
                        DailyLogEntry(
                            id = entry.id,
                            activityName = entry.activityLabel.name,
                            locationNames = entry.locationLabels.map { it.name },
                            statusName = entry.statusLabel?.name,
                            percentComplete = entry.percentComplete,
                            notes = entry.notes
                        )
                    },
                    crewMembers = log.crewMembers.map { crew ->
                        CrewMember(
                            id = crew.id,
                            name = crew.name,
                            hours = crew.hours,
                            trade = crew.trade
                        )
                    },
                    materials = log.materials.map { material ->
                        Material(
                            id = material.id,
                            name = material.name,
                            quantity = material.quantity,
                            unit = material.unit,
                            notes = material.notes
                        )
                    },
                    issues = log.issues.map { issue ->
                        Issue(
                            id = issue.id,
                            description = issue.description,
                            severity = issue.severity,
                            resolved = issue.resolved
                        )
                    },
                    photoCount = log.photoCount,
                    crewCount = log.crewCount,
                    entriesCount = log.entriesCount,
                    totalLaborHours = log.totalLaborHours
                )
            )
        } catch (e: Exception) {
            DailyLogsResult.Error(e.message ?: "Network error", isOffline = true)
        }
    }

    suspend fun createDailyLog(
        projectId: String,
        date: LocalDate,
        notes: String?,
        weatherDelay: Boolean = false,
        weatherDelayNotes: String? = null,
        gpsLatitude: Double? = null,
        gpsLongitude: Double? = null
    ): DailyLogsResult<String> = withContext(Dispatchers.IO) {
        try {
            val input = CreateDailyLogInput(
                projectId = projectId,
                date = date.toString(),
                notes = Optional.presentIfNotNull(notes),
                weatherDelay = Optional.presentIfNotNull(weatherDelay),
                weatherDelayNotes = Optional.presentIfNotNull(weatherDelayNotes),
                gpsLatitude = Optional.presentIfNotNull(gpsLatitude),
                gpsLongitude = Optional.presentIfNotNull(gpsLongitude)
            )

            val response = apolloClient.mutation(CreateDailyLogMutation(input)).execute()

            if (response.hasErrors()) {
                val errorMessage = response.errors?.firstOrNull()?.message ?: "Unknown error"
                return@withContext DailyLogsResult.Error(errorMessage)
            }

            val createdLog = response.data?.createDailyLog
                ?: return@withContext DailyLogsResult.Error("Failed to create daily log")

            DailyLogsResult.Success(createdLog.id)

        } catch (e: Exception) {
            // Queue for offline sync
            val localId = "local_${UUID.randomUUID()}"
            val payload = mapOf(
                "projectId" to projectId,
                "date" to date.toString(),
                "notes" to notes,
                "weatherDelay" to weatherDelay,
                "weatherDelayNotes" to weatherDelayNotes,
                "gpsLatitude" to gpsLatitude,
                "gpsLongitude" to gpsLongitude
            )

            pendingActionDao.upsert(
                PendingActionEntity(
                    id = localId,
                    type = "CREATE_DAILY_LOG",
                    resourceId = localId,
                    payloadJson = json.encodeToString(payload),
                    status = "pending",
                    createdAt = System.currentTimeMillis(),
                    retryCount = 0,
                    priority = 1
                )
            )

            // Also save locally
            dailyLogDao.insertAll(
                listOf(
                    DailyLogEntity(
                        id = localId,
                        date = date.toString(),
                        status = "DRAFT",
                        notes = notes,
                        projectId = projectId,
                        projectName = "", // Will be filled on sync
                        submitterName = "", // Will be filled on sync
                        crewCount = 0,
                        totalHours = 0.0,
                        pendingSync = true,
                        updatedAt = System.currentTimeMillis()
                    )
                )
            )

            DailyLogsResult.Success(localId)
        }
    }

    suspend fun updateDailyLog(
        id: String,
        notes: String?,
        weatherDelay: Boolean?,
        weatherDelayNotes: String?
    ): DailyLogsResult<Unit> = withContext(Dispatchers.IO) {
        try {
            val input = UpdateDailyLogInput(
                notes = Optional.presentIfNotNull(notes),
                weatherDelay = Optional.presentIfNotNull(weatherDelay),
                weatherDelayNotes = Optional.presentIfNotNull(weatherDelayNotes)
            )

            val response = apolloClient.mutation(UpdateDailyLogMutation(id, input)).execute()

            if (response.hasErrors()) {
                val errorMessage = response.errors?.firstOrNull()?.message ?: "Unknown error"
                return@withContext DailyLogsResult.Error(errorMessage)
            }

            DailyLogsResult.Success(Unit)

        } catch (e: Exception) {
            // Queue for offline sync
            val payload = mapOf(
                "id" to id,
                "notes" to notes,
                "weatherDelay" to weatherDelay,
                "weatherDelayNotes" to weatherDelayNotes
            )

            pendingActionDao.upsert(
                PendingActionEntity(
                    id = UUID.randomUUID().toString(),
                    type = "UPDATE_DAILY_LOG",
                    resourceId = id,
                    payloadJson = json.encodeToString(payload),
                    status = "pending",
                    createdAt = System.currentTimeMillis(),
                    retryCount = 0,
                    priority = 1
                )
            )

            DailyLogsResult.Success(Unit)
        }
    }

    suspend fun submitDailyLog(id: String): DailyLogsResult<Unit> = withContext(Dispatchers.IO) {
        try {
            val response = apolloClient.mutation(SubmitDailyLogMutation(id)).execute()

            if (response.hasErrors()) {
                val errorMessage = response.errors?.firstOrNull()?.message ?: "Unknown error"
                return@withContext DailyLogsResult.Error(errorMessage)
            }

            DailyLogsResult.Success(Unit)

        } catch (e: Exception) {
            // Queue for offline sync
            pendingActionDao.upsert(
                PendingActionEntity(
                    id = UUID.randomUUID().toString(),
                    type = "SUBMIT_DAILY_LOG",
                    resourceId = id,
                    payloadJson = """{"id": "$id"}""",
                    status = "pending",
                    createdAt = System.currentTimeMillis(),
                    retryCount = 0,
                    priority = 2
                )
            )

            DailyLogsResult.Success(Unit)
        }
    }

    private fun DailyLogEntity.toSummary() = DailyLogSummary(
        id = id,
        date = LocalDate.parse(date),
        status = status ?: "DRAFT",
        notes = notes,
        projectId = projectId,
        projectName = projectName ?: "",
        submitterName = submitterName ?: "",
        photoCount = 0, // Not stored in entity
        crewCount = crewCount ?: 0,
        entriesCount = entriesCount ?: 0,
        isPending = pendingSync
    )
}
